docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -221,24 +221,44 @@ Horizontal Pod Autoscalers also need to be considered and removed.
+**Using timestamps in the ETCD backup entries:**
 
-Currently we have come up with two ideas to handle this case:
+The leader election depends on two header files:
+1. Header file in the `Shoot`'s backup parent directory indicating who is the master and when the migration happened as a timestamp.
+2. Header file in each snapshot folder with timestamps.
 
-**DNS leader election:** A DNS TXT entry with TTL=60s and value seed='__Source Seed__' is used. The record is created and maintained by the Gardener Controller Manager (by using the DNS Controller Manager and its DNSEntry resource). When a control plane migration is detected, the Gardener Controller Manager changes the value of the DNS Entry to seed='__Destination Seed__' and waits for 2*TTL + 1 = 121 seconds to ensure that the change is propagated to all controllers in the old seed. We rely on the fact that DNS is highly available (100% for AWS Route53) and that control plane components in the __Source Seed__ can see the changes.
+When migration is triggered:
+1. The ETCD Druid in the __Destination Seed__ is notified of the start of the migration. This "notification" can happen via the `Cluster` resource with an additional field indicating when the migration was actually triggered as a timestamp.
+```
+spec:
+  migrationTimestamp: "2018-11-18T15:14:57Z"
+```
+2. The ETCD Druid adds this migration timestamp to the header file inside the `Shoot`'s backup parent directory together with the __Destination Seed's__ name to indicate who the master is.
+3. When the `Shoot`'s ETCD pod is launched in the __Destination Seed__ the backup-restore sidecar checks the available backups. It restores the ETCD data from the snapshot with the latest timestamp (older then the migration timestamp).
+4. The `Shoot`'s ETCD `backup-restore` sidecar in the __Source Seed__ checks if the `Seed` name in the backup parent directory header matches the current __Source Seed__.
+    - If it does not match, it does not take a snapshot and should kill the ETCD container.
+    - If it matches - migration has not been triggerd a snapshot can be made.
+5. Once the ETCD container dies, all components which depend on it (e.g. the Kube APIServer,  Machine Controller Manager and so on) will also stop working.
+
+However, additional work has to be done for other components which do not rely on the `Shoot`'s ETCD. Following is one proposal to handle this case:
+
+The `Cluster` resource has an additional timestamp field which is updated either by the ETCD Druid or the `backup-restore` sidecar every `X` seconds. If the ETCD Druid is used to update the timestamp it must monitor the health of the `backup-restore` sidecar. However, if `backup-restore` is used - it must have knowledge of kubernetes.
+```
+spec:
+  leadershipTimestamp: "2020-10-29T15:14:57Z"
+```
+Before extension controllers reconcile their custom resources they check this timestamp in the `Cluster` resource to see whether the current `Seed` is the leader:
```

Comment:
Can you explain why this is necessary? Extension controllers only reconcile when the gardenlet demands it (by annotating their custom resource with `gardener.cloud/operation=reconcile`). However, if the gardenlet already knows that it is not responsible then it won't annotate in the first place.
Okay, while writing this: It can happen that extension controllers are in an exponential backoff and continue to reconcile their resources (despite the fact that the migration was triggered in the mean time and their seed is no longer the leader).

Anyways, one step back. This whole proposals is about the scenario where the source seed is unavailable/broken/down for whatever reason. Most likely, the etcd/kube-apiserver of this source seed will be offline in this case. Hence, updating the `Cluster` (or any other resource) won't be possible in the first place, right?
OK, this might still be necessary when the source seed comes up again. 

Well, can you update this document with such details? It wasn't clear in the first place but only when I started writing about it.

Reply:
Yep, those are exactly the scenarios I wanted to cover. I'll add them to the doc.

Reply:
I agree with the approach suggested by @rfranzke. Seems simpler.

Reply:
Also, for the case where the source seed comes up again, can't we make the `gardenlet` update the ownership information in the `Cluster` resouce spec (in addition to the `Etcd` resource spec) which then can be used by the extensions. This will eliminate the need to lease/leadershipTimestamp from druid or the backup-restore sidecar and will also be a lot less stressful to the _Source Seed_ apiserver.

This approach is simpler and workable because in the migration scenario (even when the source seed is down) we only need leader _detection_ because leader _election_ is already implicitly done centrally in the garden cluster when the destination seed was chosen for the shoot and the migration was triggered. I think a true leader election based on leader timestamp or lease unnecessary (and sub-optimal) here.

Reply:
One problem here would be if the source seed's APIServer comes back up but there is still no connectivity to the garden cluster. Then the gardenlet wouldn't know who the leader is. 

We have to decide what extension controllers do in that case:
If migration has not been triggered and the extension controllers have resources to reconcile due to exponential backoff, should they reconcile them when there is no connectivity to the garden cluster?
If we want the extension controllers to reconcile their resources then we should depend on a lease object or the etcd resource as that is the only reliable way for them to know that this seed is still master.